File: hyperLocal-multivendor-eCommerce-backend-admin-seller-v1.0.0/app/Models/Store.php
<?php

namespace App\Models;

use App\Enums\SpatieMediaCollectionName;
use App\Enums\Store\StoreFulfillmentTypeEnum;
use App\Enums\Store\StoreStatusEnum;
use App\Enums\Store\StoreVerificationStatusEnum;
use App\Enums\Store\StoreVisibilityStatusEnum;
use App\Services\DeliveryZoneService;
use Carbon\Carbon;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Support\Str;
use Spatie\MediaLibrary\HasMedia;
use Spatie\MediaLibrary\InteractsWithMedia;

class Store extends Model implements HasMedia
{
    use SoftDeletes, InteractsWithMedia;

    protected $appends = ['distance'];
    protected $fillable = [
        'seller_id',
        'name',
        'slug',
        'address',
        'city',
        'landmark',
        'state',
        'zipcode',
        'country',
        'country_code',
        'latitude',
        'longitude',
        'contact_email',
        'contact_number',
        'description',
        'timing',
        'address_proof',
        'voided_check',
        'tax_name',
        'tax_number',
        'bank_name',
        'bank_branch_code',
        'account_holder_name',
        'account_number',
        'routing_number',
        'bank_account_type',
        'currency_code',
        'status',
        'max_delivery_distance',
        'order_preparation_time',
        'promotional_text',
        'about_us',
        'return_replacement_policy',
        'refund_policy',
        'terms_and_conditions',
        'delivery_policy',
        'domestic_shipping_charges',
        'international_shipping_charges',
        'metadata',
        'verification_status',
        'visibility_status',
        'fulfillment_type'
    ];

    protected $casts = [
        'metadata' => 'array',
        'max_delivery_distance' => 'double',
        'domestic_shipping_charges' => 'decimal:2',
        'international_shipping_charges' => 'decimal:2',
        'fulfillment_type' => StoreFulfillmentTypeEnum::class,
//        'status' => StoreStatusEnum::class,
        'verification_status' => StoreVerificationStatusEnum::class,
        'visibility_status' => StoreVisibilityStatusEnum::class,
    ];

    public function setNameAttribute($value): void
    {
        $this->attributes['name'] = $value;
        $this->attributes['slug'] = generateUniqueSlug(self::class, $value);
    }

    public function getDistanceAttribute()
    {
        if (!isset($this->user_latitude) || !isset($this->user_longitude)) {
            return null;
        }

        return DeliveryZoneService::calculateDistance(
            $this->user_latitude,
            $this->user_longitude,
            $this->latitude,
            $this->longitude
        );
    }

    public function seller(): BelongsTo
    {
        return $this->belongsTo(Seller::class);
    }

    public function reviews(): HasMany
    {
        return $this->hasMany(Review::class);
    }

    public function productVariants(): HasMany
    {
        return $this->hasMany(StoreProductVariant::class);
    }

    public function inventoryLogs(): HasMany
    {
        return $this->hasMany(StoreInventoryLog::class);
    }

    public function feedbacks(): HasMany
    {
        return $this->hasMany(SellerFeedback::class, 'store_id');
    }

    public function zones(): BelongsToMany
    {
        return $this->belongsToMany(DeliveryZone::class, 'store_zone', 'store_id', 'zone_id');
    }

    public function getProductCountAttribute(): int
    {
        $storeId = $this->id ?? 0;
        return Product::whereHas('variants.storeProductVariants', function ($q) use ($storeId) {
            $q->where('store_id', $storeId);
        })->count();
    }

    // get store all product avg rating
    public function getAvgProductsRatingAttribute()
    {
        return Review::whereHas('product.variants.storeProductVariants', function ($q) {
            $q->where('store_id', $this->id);
        })
            ->avg('rating');
    }

    public function getStoreLogoAttribute(): ?string
    {
        return $this->getFirstMediaUrl(SpatieMediaCollectionName::STORE_LOGO()) ?? null;
    }

    public function getStoreBannerAttribute(): ?string
    {
        return $this->getFirstMediaUrl(SpatieMediaCollectionName::STORE_BANNER()) ?? null;
    }

    public function getAddressProofAttribute(): ?string
    {
        return $this->getFirstMediaUrl(SpatieMediaCollectionName::ADDRESS_PROOF()) ?? null;
    }

    public function getVoidedCheckAttribute(): ?string
    {
        return $this->getFirstMediaUrl(SpatieMediaCollectionName::VOIDED_CHECK()) ?? null;
    }

    /**
     * Check store status based on status field
     */
    public function checkStoreStatus(): array
    {
        $isOpen = $this->status === StoreStatusEnum::ONLINE();

        return [
            'is_open' => $isOpen,
            'status' => $this->status,
        ];
    }

    /**
     * Check if store is online
     */
    public function isOnline(): bool
    {
        return $this->status === StoreStatusEnum::ONLINE();
    }

    /**
     * Check if store is offline
     */
    public function isOffline(): bool
    {
        return $this->status === StoreStatusEnum::OFFLINE();
    }

    public function registerMediaCollections(): void
    {
        $this->addMediaCollection(SpatieMediaCollectionName::STORE_LOGO())->singleFile();
        $this->addMediaCollection(SpatieMediaCollectionName::STORE_BANNER())->singleFile();
    }
}

File: backend/src/models/store.rs
use chrono::NaiveDateTime;
use diesel::prelude::*;
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;

use crate::schema::stores;

#[derive(Debug, Clone, Queryable, Selectable, Serialize, Deserialize, Identifiable)]
#[diesel(table_name = stores)]
#[diesel(check_for_backend(diesel::mysql::Mysql))]
pub struct Store {
    pub id: u64,
    pub seller_id: u64,
    pub delivery_zone_id: Option<u64>,
    pub name: String,
    pub slug: String,
    pub address: String,
    pub city: String,
    pub landmark: String,
    pub state: String,
    pub zipcode: String,
    pub country: String,
    pub country_code: String,
    pub latitude: Decimal,
    pub longitude: Decimal,
    pub contact_email: String,
    pub contact_number: String,
    pub description: Option<String>,
    pub store_url: Option<String>,
    pub timing: Option<String>,
    pub address_proof: String,
    pub voided_check: String,
    pub tax_name: String,
    pub tax_number: String,
    pub bank_name: String,
    pub bank_branch_code: String,
    pub account_holder_name: String,
    pub account_number: String,
    pub routing_number: String,
    pub bank_account_type: String,
    pub currency_code: String,
    pub permissions: Option<String>,
    pub time_slot_config: Option<JsonValue>,
    pub max_delivery_distance: f64,
    pub shipping_min_free_delivery_amount: f64,
    pub shipping_charge_priority: Option<String>,
    pub allowed_order_per_time_slot: Option<i32>,
    pub order_preparation_time: i32,
    pub carrier_partner: Option<String>,
    pub promotional_text: Option<String>,
    pub about_us: String,
    pub return_replacement_policy: Option<String>,
    pub refund_policy: Option<String>,
    pub terms_and_conditions: Option<String>,
    pub delivery_policy: Option<String>,
    pub shipping_preference: Option<String>,
    pub domestic_shipping_charges: Option<Decimal>,
    pub international_shipping_charges: Option<Decimal>,
    pub metadata: JsonValue,
    pub verification_status: String,
    pub visibility_status: String,
    pub fulfillment_type: String,
    pub status: Option<String>,
    pub deleted_at: Option<NaiveDateTime>,
    pub created_at: Option<NaiveDateTime>,
    pub updated_at: Option<NaiveDateTime>,
}

#[derive(Debug, Clone, AsChangeset, Deserialize)]
#[diesel(table_name = stores)]
pub struct UpdateStore {
    pub name: Option<String>,
    pub address: Option<String>,
    pub city: Option<String>,
    pub state: Option<String>,
    pub verification_status: Option<String>,
    pub visibility_status: Option<String>,
    pub status: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct StoreResponse {
    pub id: u64,
    pub seller_id: u64,
    pub name: String,
    pub slug: String,
    pub city: String,
    pub state: String,
    pub country: String,
    pub contact_email: String,
    pub contact_number: String,
    pub verification_status: String,
    pub visibility_status: String,
    pub fulfillment_type: String,
    pub created_at: Option<NaiveDateTime>,
}

impl From<Store> for StoreResponse {
    fn from(s: Store) -> Self {
        Self {
            id: s.id,
            seller_id: s.seller_id,
            name: s.name,
            slug: s.slug,
            city: s.city,
            state: s.state,
            country: s.country,
            contact_email: s.contact_email,
            contact_number: s.contact_number,
            verification_status: s.verification_status,
            visibility_status: s.visibility_status,
            fulfillment_type: s.fulfillment_type,
            created_at: s.created_at,
        }
    }
}
